nrf52-hal-common/src/spim.rs

```diff
@@ -119,79 +185,64 @@ impl<T> Spim<T> where T: SpimExt {
         Spim(spim)
     }
 
-    /// Read from an SPI slave
-    ///
-    /// This method implements a complete read transaction, which consists of
-    /// the master transmitting what it wishes to read, and the slave responding
-    /// with the requested data.
-    ///
-    /// Uses the provided chip select pin to initiate the transaction. Transmits
-    /// all bytes in `tx_buffer`, then receives bytes until `rx_buffer` is full.
-    /// Both buffer must have a length of at most 255 bytes.
-    pub fn read(&mut self,
-        chip_select: &mut P0_Pin<Output<PushPull>>,
-        tx_buffer  : &[u8],
-        rx_buffer  : &mut [u8],
-    )
-        -> Result<(), Error>
+    /// Internal helper function to setup and execute SPIM DMA transfer
+    fn  do_spi_dma_transfer<CSFun>(&mut self,
+            tx_data_ptr:u32,
+            tx_len:u32,
+            rx_data_ptr:u32,
+            rx_len:u32,
+            mut cs_n:CSFun
+            ) -> Result<(), Error>
+            where CSFun: FnMut(bool)
     {
-        if tx_buffer.len() > EASY_DMA_SIZE {
-            return Err(Error::TxBufferTooLong);
+        // Check If buffer is in data RAM, compiler sometimes put static data
+        // in flash this area is not accessable by EasyDMA
+        if (tx_data_ptr as usize) < SRAM_LOWER || tx_data_ptr as usize >= SRAM_UPPER {
```

Comment:
If all the methods calling this one made sure to copy to SRAM, if needed, we could replace this with an `assert!` and remove the new error variants again.

Reply:
I intended this check to be propagated to all the outer (API) functions calling this. I placed it in the  inner function so the code would be shared.

There are still cases where this code can be called with buffer in flash which is not handled, when all cases are handled correctly this could be an assert until then I think it is an error.